Remove template related code and update language versions

dinocosta opened the pull request:
This Pull Request introduces the following changes to this library:

* Remove code related to the reCAPTCHA widget as we're only using the API Verification request
    * This also entails updating the project's README file, removing any mention of the widget
* Update mix formatter settings file to match what we're using on Tiger
* Update Elixir and Erlang version to match what we're using on Tiger
* Update all files under `config/` to use `import Config` instead of the deprecated `use Mix.Config` pattern
* Fix test related to multiple errors being returned by Google's reCAPTCHA API
    * I was not able to get the API to return multiple errors so I ended up adding another clause to the HTTP Mock Client in order to test this and get the test to pass
  
 Lastly, I was able to successfully test this changes locally by enabling the reCAPTCHA verification in Tiger and updating its dependencies to point to a local copy of this branch:
 
```elixir
defmodule Tiger.MixProject do
  ...
  
  defp deps do
    [
      {:recaptcha, path: "/Users/dino/Developer/recaptcha"},
      ...
    ]
  end
  
  ...
end
```

dinocosta committed:
refactor: remove template code

Remove any code related to rendering the reCAPTCHA widget as we'll only
need the code to send a verification request to reCATPCHA's API.

dinocosta committed:
feat: update Elixir and Erlang versions

Update ther versions of Elixir and Erlang defined in the .tool-versions
file. Also update one of the project's dependencies to a more recent
version.

dinocosta committed:
fix: fix failing test

dinocosta committed:
refactor: update mix formatter settings

Update the .formatter.exs settings to match what we are already using in
other Elixir codebases.

vorce commented:
Taking a look at this tomorrow or Monday!

vorce commented on lib/recaptcha/http/mock_http_client.ex:
```diff
@@ -24,14 +23,21 @@ defmodule Recaptcha.Http.MockClient do
     {:error, [:"invalid-input-response"]}
   end
 
+  # This clause is used to simulate reCAPTCHA's API returning multiple error codes as the old test
+  # wasn't working anymore and I was not able to make the API trigger multiple error codes manually,
+  # even though the `error-codes` field in the response is a list of errors.
+  def request_verification("response=not_valid&secret=test_secret" = body, options) do
+    send(self(), {:request_verification, body, options})
+    {:error, [:invalid_input_response, :invalid_input_secret]}
+  end
```
Maybe it's worth changing the tests to use [bypass](https://hex.pm/packages/bypass) at some point 🤔

dinocosta replied:
@vorce Thank you so much for that recommendation! I had never used bypass before and it was awesome to get the chance to learn about it and use it 🙌 

I have pushed [this commit](https://github.com/remoteoss/recaptcha/pull/1/commits/d1b9527ee0c7817fb6eaa83814f3b615db0e2769) which introduces bypass a project dependency and uses it in the tests, please give it a review as it's my first time using it.

I believe it's also worth noting that I didn't delete the `Recaptcha.Http.MockClient` module as the tests are using the `assert_received` call in order to validate the arguments provided to the `request_verification/2` function. I did update it so that it only sends the message, with `send/2`, but then calls the real HTTP client. Do you think it's worth it to introduce Mox in order to perform this argument validation and completely remove this module? 🤔

vorce replied:
Wow, nice one @joaofcosta ! Looks great 👍 

> Do you think it's worth it to introduce Mox in order to perform this argument validation and completely remove this module?

Nah, unless it bothers you hehe. I think what we got now is fine!

vorce commented on README.md:
```diff
@@ -12,16 +12,6 @@ A simple Elixir package for implementing [reCAPTCHA] in Elixir applications.
 
 [reCAPTCHA]: http://www.google.com/recaptcha
 
-## Migration from 1.x
-
-### Breaking Changes
-
-1. Template functionality is now in a separate module: `Recaptcha.Template`. Please note: in future templating may move to a Phoenix specific package.
-2. `verify` API has changed, see the code for documentation of the new API.
-
-Most other questions about 2.x should be answered by looking over the documentation and the code. Please raise an issue
-if you have any problems with migrating.
```
Maybe we should keep this but change it to: "Migration from 1.x -> 2.x"

dinocosta replied:
@vorce Sounds good to me. Initially deleted it as we're completely moving in a different direction from the original library but I guess it makes sense to keep this section for users that do happen to be using those versions.

I have added this section back in [this commit](https://github.com/remoteoss/recaptcha/pull/1/commits/947456c36cd8abc12e632db4bb809a94fca8a10b) 🙂

vorce reviewed (approved):
We should bump the major version as well. Since we're breaking compatibility

dinocosta committed:
refactor: update major version

Update the version from 3.1.0 to 4.0.0 as removing the ability to render
the reCAPTCHA widget is a breaking change.

dinocosta commented:
> We should bump the major version as well. Since we're breaking compatibility

@vorce You're totally right, I have updated it from 3.1.0 to 4.0.0 with [this commit](https://github.com/remoteoss/recaptcha/pull/1/commits/8b0c3e3b9cfe2d319ec57e102f29f99656994e15). Should I also update the elixir version from 1.6 to 1.13 as that's the version on the `.tool-versions` file in this PR? 🤔

dinocosta committed:
refactor: keep section about migration in README

Even though we are now on version 4.0.0 of this library it might still
be worth it to keep a section about migration from 1.X to 2.X if someone
happens to decide to use those versions.

dinocosta committed:
refactor: use bypass in tests

This commit introduces bypass a project dependency when in the test
environment. Instead of relying in the `Recaptcha.Http.MockClient`
module to return a pre-defined response depending on the route we are
instead using bypass to return HTTP responses mimicking what is expected
from Google's reCATPCHA API.

vorce commented:
> > We should bump the major version as well. Since we're breaking compatibility
> 
> @vorce You're totally right, I have updated it from 3.1.0 to 4.0.0 with [this commit](https://github.com/remoteoss/recaptcha/pull/1/commits/8b0c3e3b9cfe2d319ec57e102f29f99656994e15). Should I also update the elixir version from 1.6 to 1.13 as that's the version on the `.tool-versions` file in this PR? 🤔

Yeah let's do that, since we're updating the library major version anyway.

vorce commented on README.md:
```diff
@@ -12,15 +12,15 @@ A simple Elixir package for implementing [reCAPTCHA] in Elixir applications.
 [reCAPTCHA]: http://www.google.com/recaptcha
 
-## Migration from 1.x
+## Migration from 1.x to 2.x
```
Should we have a short 4.x section?

```
## Migration from 3.x to 4.x

- Now requires elixir version 1.13 or later
- The recaptcha widget has been removed. If you need this we suggest you inline it in your project from the previous version of this library.
```

dinocosta replied:
Nice suggestion! I have pushed [this commit](https://github.com/remoteoss/recaptcha/pull/1/commits/ee1adb243dbb6940684808aac8e768c7159ea8e9) which adds that section, removes some of the badges as they were pointing to the original repo and updates the installation instructions, as our version of the library is not on hex.pm . Let me know what you think 🙂

vorce replied:
Excellent! ✨

dinocosta committed:
docs: improve README

- Add section about migrating from 3.X to 4.X
- Update installation section by using the `:github` option, as this
  library is not available on hex.pm.
- Remove badges that were still pointing to the original library.

dinocosta committed:
refactor: update elixir version in mix.exs

dinocosta commented:
> > > We should bump the major version as well. Since we're breaking compatibility
> > 
> > 
> > @vorce You're totally right, I have updated it from 3.1.0 to 4.0.0 with [this commit](https://github.com/remoteoss/recaptcha/pull/1/commits/8b0c3e3b9cfe2d319ec57e102f29f99656994e15). Should I also update the elixir version from 1.6 to 1.13 as that's the version on the `.tool-versions` file in this PR? 🤔
> 
> Yeah let's do that, since we're updating the library major version anyway.

@vorce Done in [this commit](https://github.com/remoteoss/recaptcha/pull/1/commits/3caf349fb72a413fe39bb996e267f281157985d3) 👍

vorce commented:
:shipit:
